File: tools/binding_compliance/conformance/families/file_backups.py
"""Managed game-file backups preserve exact copies, restoration and removal."""

from collections.abc import Mapping

from ..coverage import CoveragePredicate, FamilyCoveragePolicy


def matches_backup(value: Mapping) -> bool:
    """Require an existing native backup and restoration of overwritten source bytes."""
    return (
        set(value)
        == {"initial", "exists", "created", "restored", "copy", "source", "files"}
        and value["initial"] is False
        and value["exists"] is True
        and value["created"] == "Backed up 1 files"
        and value["restored"] == 1
        and value["copy"] == "original bytes\n"
        and value["source"] == value["copy"]
        and value["files"]
        == {
            "f4se_fixture.dll": value["copy"],
            "sentinel.txt": "keep\n",
            "CLASSIC_Backups/XSE_Backup/f4se_fixture.dll": value["copy"],
        }
    )


def matches_removed(value: Mapping) -> bool:
    """Backup removal must preserve the restored game file and unrelated sentinel."""
    return value == {
        "exists": False,
        "files": {"f4se_fixture.dll": "original bytes\n", "sentinel.txt": "keep\n"},
    }


def matches_game_files(value: Mapping) -> bool:
    """Require all three public operations with intermediate restored source bytes."""
    return value == {
        "backup": "1 files affected, 0 errors",
        "restore": "1 files affected, 0 errors",
        "remove": "1 files affected, 0 errors",
        "restored": "original bytes\n",
        "files": {
            "backups/fixture/f4se_fixture.dll": "original bytes\n",
            "game/sentinel.txt": "keep\n",
        },
    }


FILE_BACKUPS_COVERAGE_POLICY = FamilyCoveragePolicy(
    "file-backups",
    (
        CoveragePredicate(
            "file-backups.managed",
            "file-backups.managed",
            "file-backups.managed",
            "durable-effects",
            ("BackupManager", "create_backup", "restore_backup", "backup_exists"),
            matches_backup,
            runtime_operations=(
                None,
                "backup_manager_new",
                "backup_manager_create",
                "backup_manager_restore",
                "backup_manager_exists",
            ),
        ),
        CoveragePredicate(
            "file-backups.remove",
            "file-backups.remove",
            "file-backups.remove",
            "durable-effects",
            ("remove_backup",),
            matches_removed,
            runtime_operations=("backup_manager_remove",),
        ),
        CoveragePredicate(
            "file-backups.game-files",
            "file-backups.game-files",
            "file-backups.game-files",
            "durable-effects",
            ("GameFilesManager", "backup", "restore", "remove"),
            matches_game_files,
            runtime_operations=(
                None,
                "game_files_manager_new",
                "game_files_backup",
                "game_files_restore",
                "game_files_remove",
            ),
        ),
    ),
)

File: business-logic/classic-scanlog-core/tests/semantic_conformance/file_backups.rs
//! Managed backup lifecycles observed through public core APIs.

use super::{RunnerResult, text};
use classic_file_io_core::{
    GameFilesManager,
    backup::{BackupManager, BackupType},
};
use serde_json::{Value, json};
use std::{fs, path::Path};

/// Retain every owned file after handles close.
fn inventory(
    root: &Path,
    directory: &Path,
    result: &mut serde_json::Map<String, Value>,
) -> RunnerResult<()> {
    for entry in fs::read_dir(directory)? {
        let path = entry?.path();
        if path.is_dir() {
            inventory(root, &path, result)?;
        } else {
            result.insert(
                path.strip_prefix(root)?
                    .to_string_lossy()
                    .replace('\\', "/"),
                json!(fs::read_to_string(path)?),
            );
        }
    }
    Ok(())
}

/// Use the shared runtime to back up, overwrite and restore owned fixture files.
pub(super) fn execute(fixture: &Value) -> RunnerResult<Value> {
    classic_shared_core::get_runtime().block_on(async {
        let temporary = tempfile::tempdir()?;
        let root = temporary.path();
        let kind = text(&fixture["kind"])?;
        let game = if kind == "game-files" { root.join("game") } else { root.to_path_buf() };
        fs::create_dir_all(&game)?;
        let source = game.join("f4se_fixture.dll");
        fs::write(&source, text(&fixture["content"])?)?;
        fs::write(game.join("sentinel.txt"), "keep\n")?;
        let mut files = serde_json::Map::new();
        if kind != "game-files" {
            let manager = BackupManager::new(game, None);
            let initial = manager.backup_exists(BackupType::XSE).await?;
            let info = manager.create_backup(BackupType::XSE).await?;
            let exists = manager.backup_exists(BackupType::XSE).await?;
            let copy = fs::read_to_string(info.backup_dir.join("f4se_fixture.dll"))?;
            fs::write(&source, "changed\n")?;
            let restored = manager.restore_backup(BackupType::XSE).await?;
            if kind == "remove" {
                manager.remove_backup(BackupType::XSE).await?;
                inventory(root, root, &mut files)?;
                return Ok(json!({"exists":manager.backup_exists(BackupType::XSE).await?,"files":files}));
            }
            inventory(root, root, &mut files)?;
            return Ok(json!({"initial":initial,"exists":exists,"created":format!("Backed up {} files",info.file_count),"restored":restored,"copy":copy,"source":fs::read_to_string(source)?,"files":files}));
        }
        let manager = GameFilesManager::new(game, root.join("backups"));
        let patterns = vec!["f4se_".to_owned()];
        let summarize = |value: classic_file_io_core::game_files::FileOperationResult| format!("{} files affected, {} errors", value.files_affected, value.errors.len());
        let backup = summarize(manager.backup("fixture", &patterns).await?);
        fs::write(&source, "changed\n")?;
        let restore = summarize(manager.restore("fixture", &patterns).await?);
        let restored = fs::read_to_string(source)?;
        let remove = summarize(manager.remove("fixture", &patterns).await?);
        inventory(root, root, &mut files)?;
        Ok(json!({"backup":backup,"restore":restore,"remove":remove,"restored":restored,"files":files}))
    })
}
